clisops/core/regrid.py

```diff
@@ -0,0 +1,1861 @@
+    # todo: should nominal_resolution of the target grid be calculated if not specified in the attr?
+    nom_res_n = grid_out.ds.attrs.pop("nominal_resolution", None)
+    if nom_res_n:
+        attrs_append["nominal_resolution"] = nom_res_n
+
+    # Remove all unnecessary coords, data_vars (and optionally attrs) from grid_out.ds
+    if keep_attrs == "target":
+        grid_out._drop_vars(keep_attrs=True)
+    else:
+        grid_out._drop_vars(keep_attrs=False)
+
+    # todo: It might in general be sufficient to always act as if the threshold was
+    #  set correctly and let xesmf handle it. But then we might not allow it for
+    #  the bilinear method, as the results do not look too great and I am still
+    #  not sure/convinced adaptive_masking makes sense for this method.
+
+    # Allow Dataset and DataArray as input, but always return a Dataset
+    if isinstance(grid_in.ds, xr.Dataset):
+        for data_var in grid_in.ds.data_vars:
+            if not all(
+                [
+                    dim in grid_in.ds[data_var].dims
+                    for dim in grid_in.ds[grid_in.lat].dims
+                    + grid_in.ds[grid_in.lon].dims
+                ]
+            ):
+                continue
+            if weights.regridder.method in [
+                "conservative",
+                "conservative_normed",
+                "patch",
+            ]:
+                # Re-normalize at least contributions from duplicated cells, if adaptive masking is deactivated
+                if (
+                    adaptive_masking_threshold < 0 or adaptive_masking_threshold > 1
+                ) and grid_in.contains_duplicated_cells:
+                    adaptive_masking_threshold = 0.0
+                grid_out.ds[data_var] = weights.regridder(
+                    grid_in.ds[data_var],
+                    skipna=True,
+                    na_thres=adaptive_masking_threshold,
+                )
+            else:
+                grid_out.ds[data_var] = weights.regridder(
+                    grid_in.ds[data_var], skipna=False
+                )
+            if keep_attrs:
+                grid_out.ds[data_var].attrs.update(grid_in.ds[data_var].attrs)
+                grid_out.ds[data_var].encoding.update(grid_in.ds[data_var].encoding)
+    else:
+        if (
+            weights.regridder.method in ["conservative", "conservative_normed", "patch"]
+            and 0.0 <= adaptive_masking_threshold <= 1.0
+        ):
+            grid_out.ds[grid_in.ds.name] = weights.regridder(
+                grid_in.ds, skipna=True, na_thres=adaptive_masking_threshold
+            )
+        else:
+            grid_out.ds[grid_in.ds.name] = weights.regridder(grid_in.ds, skipna=False)
+        if keep_attrs:
+            # FIXME: This will fail - data_var is undefined.
+            grid_out.ds[data_var].attrs.update(grid_in.ds[data_var].attrs)
+            grid_out.ds[data_var].encoding.update(grid_in.ds[data_var].encoding)
```

Comment:
This should be addressed.

Reply:
I noticed there was no test to regrid a DataArray. I added one and there were a bunch of problems to be fixed.